lib/rmg/CRmgTemplateZone.cpp

```diff
@@ -243,6 +262,481 @@ void CRmgTemplateZone::createBorder()
+			}
+		}
+	}
+	
+	std::list<int3> tilesQueue(waterTiles.begin(), waterTiles.end()); //tiles need to be processed
+	std::set<int3> tilesChecked = waterTiles; //tiles already processed
+	std::map<int, std::set<int3>> coastTilesMap; //key: distance to water; value: tiles with that distance
+	std::map<int3, int> tilesDist; //key: tile; value: distance to water
+	
+	//optimization: prefill distance for all tiles marked for water with 0
+	for(auto& tile : waterTiles)
+	{
+		tilesDist[tile] = 0;
+	}
+	
+	//lambda for increasing distance of negihbour tiles
+	auto coastSearch = [this, &tilesDist, &tilesChecked, &coastTilesMap, &tilesQueue](const int3 & src, const int3 & dst)
+	{
+		if(!gen->map->isInTheMap(dst))
+			return;
+		
+		if(tilesChecked.find(dst)!=tilesChecked.end())
+			return;
+		
+		if(tileinfo.find(dst) != tileinfo.end())
+		{
+			tilesDist[dst] = tilesDist[src] + 1;
+			coastTilesMap[tilesDist[dst]].insert(dst);
+			tilesChecked.insert(dst);
+			tilesQueue.push_back(dst);
+		}
+	};
+	
+	//fills the distance-to-water map
+	while(!tilesQueue.empty())
+	{
+		int3 tile = tilesQueue.front();
+		tilesQueue.pop_front();
+		gen->foreachDirectNeighbour(tile, std::bind(coastSearch, tile, std::placeholders::_1));
+	}
+	
+#ifdef _DEBUG
+	{
+		static int _distId = 0;
+		std::ofstream out("water_dist"+std::to_string(_distId++)+".txt");
+		int levels = gen->map->twoLevel ? 2 : 1;
+		int width =  gen->map->width;
+		int height = gen->map->height;
+		for (int k = 0; k < levels; k++)
+		{
+			for(int j=0; j<height; j++)
+			{
+				for (int i=0; i<width; i++)
+				{
+					int3 tile{i,j,k};
+					int d = tilesDist[tile];
+					if(d == 0)
+						out << '~';
+					if(d>9)
+						out << '#';
+					if(d>0 && d<=9)
+						out << d;
+				}
+				out << std::endl;
+			}
+			out << std::endl;
+		}
+		out << std::endl;
+	}
+#endif
+	
+	//lambda which finds all neighbour tiles need to became water
+	auto coastPlacer = [this, &tilesDist, &tilesChecked, &tilesQueue](const int3 & src, const int3 & dst)
+	{
+		if(!gen->map->isInTheMap(dst))
+			return;
+		
+		if(tilesChecked.find(dst)!=tilesChecked.end())
+			return;
+		
+		if(tilesDist[dst] > 0 && tilesDist[src]-tilesDist[dst] == 1)
+		{
+			tilesQueue.push_back(dst);
+			tilesChecked.insert(dst);
+		}
+	};
+	
+	
+	//generating some irregularity of coast
+	int coastIdMax = fmin(sqrt(coastTilesMap.size()), 7.f); //size of coastTilesMap shows the most distant tile from water
+	assert(coastIdMax>0);
+	tilesChecked.clear();
+	for(int coastId=coastIdMax; coastId>=1; --coastId)
+	{
+		//amount of iterations shall be proportion of coast perimeter
+		const int coastLength = coastTilesMap[coastId].size() / (coastId+3);
+		for(int coastIter = 0; coastIter < coastLength; ++coastIter)
+		{
+			int3 tile = *RandomGeneratorUtil::nextItem(coastTilesMap[coastId], gen->rand);
+			if(tilesChecked.find(tile)!=tilesChecked.end())
+				continue;
+			if(gen->isUsed(tile) || gen->isFree(tile)) //prevent placing water nearby town
```

Comment:
I thought about feature to allow placement of some towns close to water in order to build shipyards. Maybe it could be template setting. However, scaling existing solution is tough, so I'm working on refactoring

Reply:
Some towns are unable to build shipyard at all, while Castle town has Lighthouse upgrade to shipyard that boosts havigation to all heroes.

Reply:
Maps where you could buy shipyard in town were common in heroes2. In heroes3 they are not because shipyard can be build in minority of towns (like 4 towns out of 9). This means that for the template with such setting those 4 towns would be preferred over the rest. Free access to shipyard and on water free resources that are unguarded means a really good economic boost in early game. However if a template would have on map unguarded shipyards then this would be ok, but why would one buy a shipyard in town and spent extra resources ?

Reply:
It is definitely is not a PR blocker. Just nice to have low priority thing. It is not required to address it.